aquadoggo/src/graphql/replication/context.rs

```diff
@@ -0,0 +1,216 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use crate::db::stores::StorageEntry;
+
+pub use super::aliased_author::AliasedAuthor;
+pub use super::public_key::PublicKey;
+use super::AuthorOrAlias;
+use super::Entry;
+use super::EntryAndPayload;
+use super::EntryHash;
+use super::LogId;
+use super::SequenceNumber;
+use super::SingleEntryAndPayload;
+use anyhow::{anyhow, Result};
+use async_graphql::ID;
+use lru::LruCache;
+use p2panda_rs::entry::{decode_entry, SeqNum};
+use p2panda_rs::storage_provider::traits::EntryStore;
+
+#[derive(Debug)]
+pub struct Context<ES: EntryStore<StorageEntry>> {
+    author_aliases: LruCache<ID, PublicKey>,
+    next_alias: usize,
+    entry_store: ES,
+}
+
+impl<ES: EntryStore<StorageEntry>> Context<ES> {
+    pub fn new(author_aliases_cache_size: usize, entry_store: ES) -> Self {
+        Self {
+            author_aliases: LruCache::new(author_aliases_cache_size),
+            next_alias: Default::default(),
+            entry_store,
+        }
+    }
+
+    pub fn insert_author_aliases(&mut self, public_keys: Vec<PublicKey>) -> Vec<AliasedAuthor> {
+        public_keys
+            .into_iter()
+            .map(|public_key| {
+                self.next_alias += 1;
+                self.author_aliases
+                    .put(ID(self.next_alias.to_string()), public_key.clone());
+                AliasedAuthor {
+                    public_key,
+                    alias: self.next_alias.into(),
+                }
+            })
+            .collect()
+    }
+
+    pub fn author_aliases_to_public_keys(
+        &mut self,
+        ids: Vec<ID>,
+    ) -> Result<Vec<Option<PublicKey>>> {
+        ids.into_iter()
+            .map(|id| {
+                let result = self.author_aliases.get(&id).map(|key| key.clone());
+                Ok(result)
+            })
+            .collect()
+    }
+
+    pub async fn entry_by_log_id_and_sequence<'a>(
+        &mut self,
+        log_id: LogId,
+        sequence_number: SequenceNumber,
+        author_alias: AuthorOrAlias,
+    ) -> Result<Option<SingleEntryAndPayload>> {
+        let author = self.get_author(author_alias)?;
+
+        let result = self
+            .entry_store
+            .get_entry_at_seq_num(&author.0, &log_id.0, &sequence_number.0)
+            .await?
+            .map(|entry| entry.into());
+
+        Ok(result)
+    }
+
+    pub async fn entry_by_hash<'a>(
+        &mut self,
+        hash: EntryHash,
+    ) -> Result<Option<SingleEntryAndPayload>> {
+        let result = self
+            .entry_store
+            .get_entry_by_hash(&hash.into())
+            .await?
+            .map(|entry_row| entry_row.into());
+
+        Ok(result)
+    }
+
+    pub async fn get_skiplinks<'a>(&mut self, entry: &Entry) -> Result<Vec<Entry>> {
+        let author = entry.as_ref().author();
+        let entry = decode_entry(entry.as_ref(), None)?;
```

Comment:
Is there a nicer way to do this rather than having to get the author from an `EntrySigned` and the `log_id` and `seq_num` from an `Entry`?

Reply:
I just looked at this again, and wonder why we don't change the signature for `get_certificate_pool()` to accept an entry hash as it's parameter? Then you wouldn't need author, log_id, seq_num.

Reply:
Yeah that's a nice idea!